Updated dependency management (#304)
bitwarden/contributing-docs: docs/contributing/dependencies/index.md

## Workflow

> Renovate is currently scheduled to automatically create pull requests every 2 weeks for all open
> dependencies. The goal of our dependency management process is for the teams to handle these
> dependencies in that same 2-week cadence, in order to avoid a large backlog of PRs and out-of-date
> packages building up over time.

Reviewer: What is an `open dependency`? 

```suggestion
Renovate is currently scheduled to automatically create pull requests every 2 weeks. The goal of our
dependency management process is for the teams to review and merge the opened pull requests in the
same 2-week cadence. To avoid a large backlog of PRs and out-of-date packages accumulating.
```

Author: Updated to use your suggestion in [`4176743` (#304)](https://github.com/bitwarden/contributing-docs/pull/304/commits/4176743c791a5540e87072b82009019827368f03).

:::info Major upgrades

> Major upgrades are an exception to this time frame, as these can take longer to coordinate. Ideally
> the team will have already coordinated and resolved deprecations in advance.

Reviewer: I think we can set the expectation more clearly.

```suggestion
Major upgrades are an exception to this time frame, as these can take longer to coordinate. The team
should make an effort to coordinate scheduled major updates and resolve deprecations in advance.
```

Author: Updated to use your suggestion in [4176743 (#304)](https://github.com/bitwarden/contributing-docs/pull/304/commits/4176743c791a5540e87072b82009019827368f03).

## Workflow

Renovate will automatically create pull requests every other week, aligning with the mid-point of
each sprint. This allows time for teams to assess the open PRs and adequately plan for them in the

> upcoming sprint.

Author: That is a great point.  I've made it more generic with [`e4973e1` (#304)](https://github.com/bitwarden/contributing-docs/pull/304/commits/e4973e15fa8fbd1352f4eeab007b8016a28fcb96).

keep groups as small as possible to minimize the impact and increase confidence in approving and
merging.

### Jira ticket

A Jira ticket will automatically be created for each Renovate PR that is opened. It will be assigned
to the appropriate team based on the dependency ownership.

The Jira ticket should be used to track the work through sprint planning, prioritization, review,
and testing.

> 

Reviewer: We use jira tickets to provide tracability for what goes into each release. The tickets also allows the team to estimate the upcoming work for the next sprint.

> 

Reviewer 2: I discussed this with the EMs a week or so ago. There's an overhead to this and we're reacting to the large amount of current work vs. future state, and I see this turning into being more selective for larger work items.

### Review

A typical dependency workflow involves the following steps:

1. Move the Jira ticket to In Progress.
2. Read the proposed changes.
3. Review the release notes of each dependency, for each released version between the current and
   the proposed upgrade. Identify if there are any deprecations or breaking changes affecting our
   code.
   1. For **breaking changes**, either resolve them yourself, or for major changes, coordinate with
      the other teams.
   2. For **deprecations**, create high priority Jira tickets on the affected teams' backlogs with a
      due date at least one sprint before the next scheduled major release of the dependency.
4. Verify CI status.
5. If test coverage is lacking, check out locally and manually confirm a few key areas.
6. Review the proposed code changes and approve the PR.
7. Write a Jira ticket containing testing notes for QA.
8. Merge the PR.

> 9. Assign the Jira ticket to QA.

Reviewer 2: ❓ We're QAing these now?

Reviewer: Yes, to follow our SDLC.

Reviewer 2: Also discussed, but the win is to not need to "QA" these, especially bumps that should guarantee existing functionality and nothing break.
